### PR TITLE
FHB-795 : Add E2E Seed Data Framework

### PR DESCRIPTION
**Pull request by AaronJY**
Ticket: [FHB-795](https://dfedigital.atlassian.net.mcas.ms/browse/FHB-795)

---

This PR:

- Adds a `setup` script to seed the databases with data for use by the [future] E2E tests

- Adds a `teardown` script to teardown the test data from the databases
- Adds documentation around the different folders, mechanisms and how to expand on the framework
- Adds documentation on how to run the `setup` and `teardown` scripts
- Adds documentation on how to set up your local environment to run the `setup` and `teardown` scripts
- Adds a suite of examples where every relevant table is seeded at least once for: Service Directory Db, Referral Db, Report Db
- Maps out the database models with the JS library Sequelize, an ORM similar to EF but in JS
- Adds wrappers to the CRUD (well, C only in our case) operations to be used when manipulating the database. It hides unused tables and columns from the testers and applies helper operations automatically without testers needing to do so manually including things like adding the test ID offset, prepending [E2E] to certain fields and encrypting strings for some test data (namely in the Referral Db)
- Has the ability to encrypt column data with the same algorithm and keys (need to obtain those yourself) as the web apps, allowing the testers to create mock data in plain English instead of encrypting and inserting strings themselves.
- Includes the skeletons for the real implementation code for when those get picked up.

---

Not in this PR:

- There is probably some more stuff that could be done around helpers and whatnot but it's better to discover and implement those if needed during real implementation rather than risk over-engineering this

- The pipeline integration of this framework will be done in a separate PR once this is merged

[FHB-795]: https://dfedigital.atlassian.net/browse/FHB-795?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,10 +6,14 @@
...
+│   ├── function <-- Serverless functions
 │   ├── service <-- API services
-│   ├── shared <-- Shared libraries (NuGet packages)
+│   ├── shared <-- Shared libraries
 │   └── ui <-- UI apps
-└── util <-- Utility scripts for common use cases
+├── util <-- Utility scripts for common use cases
+├── docs <-- Technical documentation
+├── terraform <-- Terraform (IaC) scripts for the Family Hubs architecture
+└── test <-- Scripts/frameworks relating to testing
```
Comment (AaronJY) [inline README.md:16]: Cheeky add: updated the readme to detail new dirs

File: .gitignore [hunk 2]
```diff
@@ -31,4 +31,6 @@ bld/
 [Oo]ut/
 msbuild.log
 msbuild.err
-msbuild.wrn
\ No newline at end of file
+msbuild.wrn
+
+local/db/sql
```
Comment (AaronJY) [inline .gitignore:36]: Cheeky add: ignore Docker data directories when using `docker compose up` from the mono repo

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+    "cSpell.words": [
+        "teardown"
```
Comment (AaronJY) [inline .vscode/settings.json:3]: It's a real word, dammit 😠

**Comment by AaronJY** [comment]
Nothing more to do on this PR for now. Whoever picks up FHB-795 (probably moi) can continue work on taking this from a proof-of-concept to a proper implementation!

File: .vscode/settings.json (hunk 3 above)
Comment (Top-Cat) [reply .vscode/settings.json:3]: teardown is a swear

File: test/e2e-seed-data-framework/.env.local [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# General
+IDS_START_FROM=1000000 # All generated test IDs start from 1,000,000
+
+# FamilyHubs.ServiceDirectory.Database
+DB_SD_HOST=localhost # Database host
```
Comment (stevesatdfe) [inline test/e2e-seed-data-framework/.env.local:5]: Do we want to constrain users to sql user/pwd? If running locally, could use integrated security. Also if we decide to proceed with the Entra work, the users will need to be Entra users. Maybe a connection string is better?

File: test/e2e-seed-data-framework/connections.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+export const serviceDirectoryDb = buildSequelizeConnection(
+  process.env.DB_SD_HOST, process.env.DB_SD_NAME, process.env.DB_SD_USER, process.env.DB_SD_PASSWORD)
+
+export const referralDb = buildSequelizeConnection(
+  process.env.DB_REFERRAL_HOST, process.env.DB_REFERRAL_NAME, process.env.DB_REFERRAL_USER, process.env.DB_REFERRAL_PASSWORD)
+
+function buildSequelizeConnection (dbHost, dbName, dbUser, dbPassword) {
+  console.log(`Building DB connection for ${dbName} @ ${dbHost} (user ${dbUser})`)
+  return new Sequelize(dbName, dbUser, dbPassword, {
```
Comment (stevesatdfe) [inline test/e2e-seed-data-framework/connections.js:12]: Do we want to test the connection by calling sequelize.authenticate(); before returning? Quick fail check.

From https://sequelize.org/docs/v6/getting-started/

File: test/e2e-seed-data-framework/scripts/setup.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  console.error('Unable to run setup:', error)
+} finally {
+  await closeConnections()
+}
+
+/**
+ * Runs the test data setup scripts.
+ */
+async function setup () {
+    console.log('Executing setup...')
```
Comment (stevesatdfe) [inline test/e2e-seed-data-framework/scripts/setup.js:17]: Over indenting aside - does this framework provide a logging api at all rather than relying on console.log?

**Review by stevesatdfe — CHANGES_REQUESTED** [review]

File: test/e2e-seed-data-framework/.env.local (hunk 4 above)
Comment (AaronJY) [reply test/e2e-seed-data-framework/.env.local:5]: Good shout. Will look to support connection strings.

File: test/e2e-seed-data-framework/connections.js (hunk 5 above)
Comment (AaronJY) [reply test/e2e-seed-data-framework/connections.js:12]: Good shout. Will call `sequelize.authenticate();`

File: test/e2e-seed-data-framework/scripts/setup.js (hunk 6 above)
Comment (AaronJY) [reply test/e2e-seed-data-framework/scripts/setup.js:17]: It doesn't, and I'd argue we don't need to add one. We really only care about making sure logs are sent to the standard output in whatever environment this runs under (console out locally, GitHub Action output in pipelines); I don't see a benefit in adding a custom logger personally

**Comment by AaronJY** [comment]
Converting this back to a draft as there are still bits to do. Whoever picks up the implementation ticket can pick this branch up and finish it off :)

File: test/e2e-seed-data-framework/.env.local (hunk 4 above)
Comment (Zac-Digital) [reply test/e2e-seed-data-framework/.env.local]: Resolved: [efca615](https://github.com/DFE-Digital/fh-services/pull/171/commits/efca6153afd4eaea42ff4d32cddbf2d4e5113377) and [ca6934c](https://github.com/DFE-Digital/fh-services/pull/171/commits/ca6934c232c6051c387b5e9ae6db3a3f69fd57ad)

File: test/e2e-seed-data-framework/connections.js (hunk 5 above)
Comment (Zac-Digital) [reply test/e2e-seed-data-framework/connections.js]: Resolved: [0b20858](https://github.com/DFE-Digital/fh-services/pull/171/commits/0b20858b688f921717d18799ba4f0bc1f84ae19b)

**Review by stevesatdfe — APPROVED** [review]
Looks ok from what I can see - certainly nothing jumping out at me. Quite a lot of code to review mind.

**Review by Top-Cat — APPROVED** [review]
